From 5bbbc21b250b2a4a4ef9d3e9461075788a19c633 Mon Sep 17 00:00:00 2001
From: Sergey Lavrentiev <s.lavrentiev@samsung.com>
Date: Mon, 17 Oct 2022 21:41:50 +0300
Subject: [PATCH 048/161] [drivers/dax/axdimm] Fix process release stage

Actual process exit status verification was causing
group_leader resource release on exit_signal = -1,
which seem to be normal value (no signal delivered)
for device close in userspace thread context.

Resolves: AXDIMM-451.

Signed-off-by: Sergey Lavrentiev <s.lavrentiev@samsung.com>
---
 drivers/dax/axdimm_process_manager.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/drivers/dax/axdimm_process_manager.c b/drivers/dax/axdimm_process_manager.c
index 88d9ff7b4..61ade7454 100644
--- a/drivers/dax/axdimm_process_manager.c
+++ b/drivers/dax/axdimm_process_manager.c
@@ -12,12 +12,16 @@
 
 enum rank_access_type { RANK_ACCESS_RD, RANK_ACCESS_WR, RANK_ACCESS_INVAL };
 
+static inline bool is_abnormal_exit_signal(int exit_sig)
+{
+	return exit_sig & PF_SIGNALED && exit_sig != SIGCHLD && exit_sig != -1;
+}
+
 static inline bool is_abnormal_process_exit(void)
 {
-	return current->flags & PF_SIGNALED ||
-	       current->exit_state == EXIT_DEAD ||
+	return current->exit_state == EXIT_DEAD ||
 	       current->exit_state == EXIT_ZOMBIE || current->exit_code != 0 ||
-	       current->exit_signal != SIGCHLD;
+	       is_abnormal_exit_signal(current->exit_signal);
 }
 
 static inline enum rank_access_type get_rank_access_type(unsigned int cmd)
-- 
2.34.1

